### DISCUSSION TITLE
The number of repeated runs is limited

### DISCUSSION THREAD

**Discussion by GitHubpathy** [pyswmm/pyswmm#529] [Q&A] [answered]
It's strange when I use python to call pyswmm for running, it always  failed at the 4093th runing. I am sure it is not the swmm model problem or the python code problem for I try serval model and the most simply code. But when I try use another swmm library named "swmm5", the issue disappear. It looks like the number of repeated runs is limited.

**Comment by bemcdonnell** [comment]
@GitHubpathy can you include the relevant code? Sounds like something isn’t closing properly. Also what version of pyswmm and what platform are you on?

**Comment by GitHubpathy** [comment]
@bemcdonnell Thank you for response. The relevant code is attached below:

**from pyswmm import Simulation

for i in range(1, 5000):
    sim = Simulation('swmm_model_test/test_for_pyswmm.inp')
    sim.execute()
    print(i)**

The version of pyswmm is 2.0.0, the platform is PyCharm 2023.3.3(Community Edition)

> @GitHubpathy can you include the relevant code? Sounds like something isn’t closing properly. Also what version of pyswmm and what platform are you on?

**Comment by bemcdonnell** [comment, marked as answer]
Try this:

```python
from pyswmm import Simulation

for i in range(1, 5000):
    with Simulation('swmm_model_test/test_for_pyswmm.inp') as sim:
    for step in sim:
        pass
    
    print(i)
```

**Comment by timothy-holmes** [comment]
Possible enhancement?

```python
class Simulation:
    def __init__(self):
        if not self._is_in_context_manager():
            warnings.warn(warning_string, RuntimeWarning)
```

**Reply by karosc** [reply]
As someone who enjoys thorough exception handling and warning messages, I really like this idea. 👏

**Reply by bemcdonnell** [reply]
@timothy-holmes if you would be open to submitting a PR, we would appreciate the help!

**Comment by bemcdonnell** [comment]
We don't usually suggest that anyone explicitly instantiates the `Simulation` class anymore.  The issue you're seeing is a windows number of files open limit for one running piece of software, I believe.  Been a long time since i played with `sim.execute()`.  It's not that popular for pyswmm since it doesn't really give you the means to interact with the simulation. 

Does this work in your loop?  

```python
sim = Simulation('swmm_model_test/test_for_pyswmm.inp')
sim.execute()
sim.close()
```

**Comment by GitHubpathy** [comment]
@bemcdonnell 
With your code, the problem is solved. Thank you!
